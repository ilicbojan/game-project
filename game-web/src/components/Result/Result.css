.result {
  min-height: 211px;
  position: relative;
  flex: 1;
  margin: 0;
}

.result-details {
  font-weight: bold;
  font-size: 1.5rem;
}
