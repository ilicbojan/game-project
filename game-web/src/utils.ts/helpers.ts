import { ChoiceId, ChoiceItem, ChoiceName, ResultType } from "../types/game";

export const ResultToColorMap: Record<ResultType, string> = {
  [ResultType.Win]: "green",
  [ResultType.Lose]: "red",
  [ResultType.Tie]: "orange",
};

export const ChoiceToEmojiMap: Record<string, string> = {
  rock: "✊", // Rock
  paper: "✋", // Paper
  scissors: "✌️", // Scissors
  lizard: "🦎", // Lizard
  spock: "🖖", // Spock
  random: "🎲", // Random choice
  "": "❓", // Unknown choice
};

const choiceItemMap: Record<string, ChoiceItem> = {
  rock: { id: ChoiceId.Rock, name: ChoiceName.Rock },
  paper: { id: ChoiceId.Paper, name: ChoiceName.Paper },
  scissors: { id: ChoiceId.Scissors, name: ChoiceName.Scissors },
  lizard: { id: ChoiceId.Lizard, name: ChoiceName.Lizard },
  spock: { id: ChoiceId.Spock, name: ChoiceName.Spock },
};

export function getChoiceItem(name: ChoiceName): ChoiceItem {
  return choiceItemMap[name];
}

export const formatDateTime = (d: Date) =>
  `${d.toLocaleDateString()} ${d.toLocaleTimeString()}`;
