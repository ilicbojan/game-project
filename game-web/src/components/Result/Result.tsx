import React from "react";
import { ChoiceItem, PlayResponse } from "../../types/game";
import LoadingSpinner from "../LoadingSpinner/LoadingSpinner";
import "./Result.css";
import { ChoiceToEmojiMap, ResultToColorMap } from "../../utils.ts/helpers";

interface Props {
  result: PlayResponse | null;
  choices: ChoiceItem[];
  loading: boolean;
}

const Result: React.FC<Props> = ({ result, choices, loading }) => {
  const playerChoice = choices.find((c) => c.id === result?.player);
  const computerChoice = choices.find((c) => c.id === result?.computer);

  return (
    <div className="result section">
      <h3>Round Result</h3>
      {loading && <LoadingSpinner />}
      {!result && !loading && (
        <p>No result available, please choose your option</p>
      )}
      {result && (
        <div>
          <p>
            Your choice: <strong>{playerChoice?.name}</strong>{" "}
            {ChoiceToEmojiMap[playerChoice?.name || ""]}
          </p>
          <p>
            Computer&apos;s choice: <strong>{computerChoice?.name}</strong>
            {ChoiceToEmojiMap[computerChoice?.name || ""]}
          </p>
          <div className="result-details">
            Result:{" "}
            <span style={{ color: ResultToColorMap[result.result] }}>
              {result.result.toUpperCase()}
            </span>
          </div>
        </div>
      )}
    </div>
  );
};

export default Result;
