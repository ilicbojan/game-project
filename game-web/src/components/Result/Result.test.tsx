import { render, screen } from "@testing-library/react";
import Result from "./Result";
import {
  ChoiceItem,
  ChoiceName,
  PlayResponse,
  ResultType,
} from "../../types/game";
import { getChoiceItem } from "../../utils.ts/helpers";

describe("Result component", () => {
  const choicesMock: ChoiceItem[] = [
    getChoiceItem(ChoiceName.Rock),
    getChoiceItem(ChoiceName.Paper),
    getChoiceItem(ChoiceName.Scissors),
  ];

  const resultMock: PlayResponse = {
    player: 1,
    computer: 2,
    result: ResultType.Lose,
  };

  it("matches the snapshot", () => {
    const { asFragment } = render(
      <Result result={resultMock} choices={choicesMock} loading={false} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders LoadingSpinner when loading is true", () => {
    const { container } = render(
      <Result result={null} choices={choicesMock} loading={true} />
    );
    const spinner = container.querySelector(".loading-spinner");

    expect(spinner).toBeInTheDocument();
  });

  it("shows fallback message when result is null and loading is false", () => {
    render(<Result result={null} choices={choicesMock} loading={false} />);

    expect(
      screen.getByText(/No result available, please choose your option/i)
    ).toBeInTheDocument();
  });

  it("handles empty choices array gracefully", () => {
    render(<Result result={resultMock} choices={[]} loading={false} />);
    expect(screen.getByText(/Your choice:/i)).toBeInTheDocument();
    expect(screen.queryByText("rock")).not.toBeInTheDocument();
    expect(screen.getByText(/Computer's choice:/i)).toBeInTheDocument();
    expect(screen.queryByText("paper")).not.toBeInTheDocument();
  });

  it("handles invalid player or computer IDs in result", () => {
    const invalidResultMock: PlayResponse = {
      player: 99,
      computer: 100,
      result: ResultType.Lose,
    };
    render(
      <Result
        result={invalidResultMock}
        choices={choicesMock}
        loading={false}
      />
    );

    expect(screen.getByText(/Your choice:/i)).toBeInTheDocument();
    expect(screen.queryByText("rock")).not.toBeInTheDocument();
    expect(screen.getByText(/Computer's choice:/i)).toBeInTheDocument();
    expect(screen.queryByText("paper")).not.toBeInTheDocument();
  });

  it("displays result when result is provided", () => {
    render(
      <Result result={resultMock} choices={choicesMock} loading={false} />
    );

    expect(screen.getByText(/Your choice:/i)).toBeInTheDocument();
    expect(screen.getByText("rock")).toBeInTheDocument();
    expect(screen.getByText(/Computer's choice:/i)).toBeInTheDocument();
    expect(screen.getByText("paper")).toBeInTheDocument();
    expect(screen.getByText(/Result:/i)).toBeInTheDocument();
    expect(screen.getByText("LOSE")).toBeInTheDocument();
  });
});
