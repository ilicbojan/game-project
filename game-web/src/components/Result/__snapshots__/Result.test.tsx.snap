// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Result component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="result section"
  >
    <h3>
      Round Result
    </h3>
    <div>
      <p>
        Your choice: 
        <strong>
          rock
        </strong>
         ✊
      </p>
      <p>
        Computer's choice: 
        <strong>
          paper
        </strong>
        ✋
      </p>
      <div
        class="result-details"
      >
        Result: 
        <span
          style="color: red;"
        >
          LOSE
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
